src/audio_unit/delay.rs

```diff
@@ -1,32 +1,102 @@
 use crate::{ring_buffer, AudioUnit, Result};
+use anyhow::anyhow;
 use cpal::StreamConfig;
 use ringbuf::{Consumer, Producer, RingBuffer};
+use wmidi::MidiMessage;
+
+pub type DelayMs = u32;
+
+type MidiMessageHandler = Box<dyn Fn(&[MidiMessage<'static>]) -> Option<DelayMs> + Send + Sync>;
 
 pub struct Delay {
+    delay_ms: DelayMs,
+    midi_message_handler: MidiMessageHandler,
+    stream_config: StreamConfig,
     producer: Producer<f32>,
     consumer: Consumer<f32>,
 }
 
 impl Delay {
-    pub fn new(stream_config: &StreamConfig, delay_ms: u32) -> Result<Self> {
-        let delay_num_frames = (delay_ms as f32 / 1_000.0) * stream_config.sample_rate.0 as f32;
-        let delay_num_samples = delay_num_frames as usize * stream_config.channels as usize;
+    pub const MIN_DELAY_MS: DelayMs = 0;
+    pub const MAX_DELAY_MS: DelayMs = 10000;
 
-        let ring = RingBuffer::new(delay_num_samples * 2);
+    pub fn new(
+        stream_config: &StreamConfig,
+        midi_message_handler: MidiMessageHandler,
+        delay_ms: DelayMs,
+    ) -> Result<Self> {
+        Self::validate_delay(delay_ms)?;
+
+        // size the ring buffer so that it can accomodate the largest allowed delay
+        let ring = RingBuffer::new(Self::delay_num_samples(stream_config, Self::MAX_DELAY_MS) * 2);
         let (mut producer, consumer) = ring.split();
 
-        ring_buffer::write_empty_samples(&mut producer, delay_num_samples)?;
+        ring_buffer::write_empty_samples(
+            &mut producer,
+            Self::delay_num_samples(stream_config, delay_ms),
+        )?;
+
+        Ok(Self {
+            delay_ms,
+            midi_message_handler,
+            stream_config: stream_config.clone(),
+            producer,
+            consumer,
+        })
+    }
+
+    fn validate_delay(delay_ms: DelayMs) -> Result<()> {
+        if delay_ms > Self::MAX_DELAY_MS {
```

Comment:
Per other comment re MIDI messages "interface" between this vs the effect, along the lines of thinking of this as exposing "knobs", maybe that means that the effect doesn't know about the valid ranges for the knobs, it just passes in a uniform value for that knob (eg 0-1 or 0-127) and then this guy is responsible for mapping the mutated normalized knob value into its own internal state (here, delay length)? That seems like it would get rid of this smell where there's external logic that's already mapping it to be < max delay ms but then this is double-checking that